content/blog/2018/06/2018-06-15-remoting-over-message-bus.adoc

```diff
@@ -0,0 +1,85 @@
+---
+layout: post
+title: "GSoC Project Intro: Jenkins Remoting over Message Bus/Queue"
+tags:
+- plugins
+- gsoc
+author: pvtuan10
+---
+
+=== About me
+My name is Pham Vu Tuan, I am a final year undergraduate student from Singapore. This is the first time I participate in Google Summer of Code and contribute to an open-source organization. I am very excited to contribute this summer.
+
+=== Mentors
+I have GSoC mentors who help me in this project https://github.com/oleg-nenashev[Oleg Nenashev] and https://github.com/Supun94[Supun Wanniarachchi]. Besides that, I also receive great support from developers in remoting project https://github.com/dwnusbaum[Devin Nusbaum] and https://github.com/jeffret-b[Jeff Thompson].
+
+=== Overview
+Current versions of Jenkins Remoting are based on the TCP protocol. If it fails, the agent connection and the build fails as well. There are also issues with traffic prioritization and multi-agent communications, which impact Jenkins stability and scalability.
+
+This project aims to develop a plugin in order to add support of a popular message queue/bus technology (Kafka) as a fault-tolerant communication layer in Jenkins.
+
+=== Why Kafka?
+When planning for this project, we want to use traditional message queue system such as https://github.com/oleg-nenashev[ActiveMQ] or https://github.com/oleg-nenashev[RabbitMQ]. However, after some discussion, we decided to have a try with https://kafka.apache.org/[Kafka] with more suitable features with this project:
+
+* Kafka itself is not a queue like ActiveMQ or RabbitMQ, it is a distributed, replicated commit log. This helps to remove message delivery complexity we have in traditional queue system.
+* We need to support data streaming as a requirement, and Kafka is good at this aspect, which RabbitMQ is lack of.
+* Kafka is said to have a better scalability and good support from the development community.
+
+=== Current State
+The project is reaching the end of the first phase and here are things we have achieved so far:
+
+* Setup project as a set of Docker Compose components: Kafka cluster, Jenkins master (with plugin) and a custom agent (JAR).
+* Create a PoC with new command transport implementation to support Kafka, which involves of command invocation, RMI, classloading and data streaming.
+* Make neccessary changes in Remoting and Jenkins core to make them extensible for the use of this project.
+* Decide to use Kafka as a suitable final implementation.
+
+We planned to release an alpha version of this plugin by the end of this phase, but decided to move this release to the second phase because we need to wait for remoting and core patches to be released.
+
+=== Project Structure
```

Comment:
Maybe it makes sense to just keep it in "Architecture" to reduce number of sections